src/lib.rs

```diff
@@ -60,7 +64,7 @@ impl<T> Items<T> {
         }
         let mut i = 0;
         while i < self.items.len() {
-            let should_remove = self.items[i].upgrade().is_none();
+            let should_remove = self.items[i].strong_count() == 0;
```

Comment:
It is stabilized in 1.41.0. and we can not use upgrade here now because this function is called in a lock so may cause deadlock like:
1. there is a active TrackedObject.
2. this method is called with lock hold, upgrade success and create a temp Arc.
3. the active TrackedObject is dropped.
4. the temp Arc created at step 2 is dropped, since it is the last Arc, the InnerActiveObject will be dropped, which will try to get the lock, but the lock is not reentrant.

Reply:
I tried a very similar refactoring as what you are offering (if not exactly the same) and reached the exact same conclusion.

I am **very bad at reasoning** about these, so proofreading your change will take a bit of time.

Reply:
In 4, the Arc being dropped is an `Arc<T>`, not an Arc<TrackedItem<T>>, so there are no race condition.